### DISCUSSION TITLE
Interacting directly with query optimizer

### DISCUSSION THREAD

**Discussion by drin** [duckdb/duckdb#10780] [Q&A] [answered]
Hello!

I would like to prototype a query processing layer that interacts with DuckDB. For this interaction, I would like to be able to interact directly with optimizers such as [join_order_optimizer.cpp#L25](https://github.com/drin/duckdb/blob/main/src/optimizer/join_order/join_order_optimizer.cpp#L25).

I would also like to be able to eventually implement a simple interface via the main duckdb binary; which, when built using `make -j8` is generated at `build/release/duckdb`.

I was hoping someone could point me to what source files get compiled into the `duckdb` binary so that I can get started? For reference, I will be using the [substrait extension](https://duckdb.org/docs/extensions/substrait#installing-and-loading) to construct logical plans and then I want to run specific optimizations such as the join order optimizer. I realize I can make my own binary, but I'm hoping to integrate into (or reuse) the interface provided by the `duckdb` binary in case something interesting can be shared or upstreamed.

Thank you for any help!

**Comment by Mause** [comment, marked as answer]
You probably want to start in https://github.com/duckdb/duckdb/tree/main/tools/shell and https://github.com/duckdb/duckdb/tree/main/src ?

**Reply by drin** [reply]
awesome! tools/shell is exactly what I was looking for. thanks!
